src/rulesets/Base/mapreduce.jl

```diff
@@ -19,6 +19,50 @@ function rrule(::typeof(sum), x::AbstractArray{T}; dims=:) where {T<:Number}
     return y, sum_pullback
 end
 
+# Can't map over Adjoint/Transpose Vector
+function rrule(
+    config::RuleConfig{>:HasReverseMode},
+    ::typeof(sum),
+    f,
+    xs::Union{Adjoint{<:Number,<:AbstractVector},Transpose{<:Number,<:AbstractVector}};
+    kwargs...
+)
+    op = xs isa Adjoint ? adjoint : transpose
+    # since summing a vector we don't need to worry about dims which simplifies adjointing
+    vector = parent(xs)
+    y, vector_sum_pb = rrule(config, sum, f, vector; kwargs...)
+    function covector_sum_pb(ȳ)
+        s̄um, f̄, v̄ = vector_sum_pb(ȳ)
+        return s̄um, f̄, op(v̄)
+    end
+
+    return y, covector_sum_pb
+end
+
+
+function rrule(
+    config::RuleConfig{>:HasReverseMode}, ::typeof(sum), f, xs::AbstractArray; dims=:
+)
+    fx_and_pullbacks = map(x->rrule_via_ad(config, f, x), xs)
+    y = sum(first, fx_and_pullbacks; dims=dims)
+
+    pullbacks = last.(fx_and_pullbacks)
+
+    function sum_pullback(ȳ)
+        call(f, x) = f(x)  # we need to broadcast this to handle dims kwarg
+        f̄_and_x̄s = call.(pullbacks, ȳ)
+        # no point thunking as most of work is in f̄_and_x̄s which we need to compute for both
+        f̄ = if fieldcount(typeof(f)) === 0 # Then don't need to worry about derivative wrt f
```

Comment:
Not necessarily related to this PR, but it would be good to define a function for `fieldcount(typeof(f)) === 0`, e.g. `hasstructure(f)` or `iscomposite(f)`

Reply:
I am not sure that is necessarily the right check though. There could be number-like or array-like functors which do have a well-defined derivative wrt `f`, but don't have any fields. Probably not a big deal for now, but that might be something to keep in mind for general design decisions.

Reply:
true, they are very unusual though. Probably a good thing to worry about when defining @CarloLucibello 's `iscomposite`/`hasstructure` or what ever we call it, maybe `hastangent`?

Cases I can think of are functors that are also:
 -  number types defined using `primative`,
  - Something that is like `FillArrays.One()` or `OneHot` that pushes  size and index into the type-level.

Reply:
@tkf's suggestion was to use `Base.issingletontype`, which I think would avoid your `primitive` example. Not sure there's a One, but other cases with a value in the type are regarded as constants:
```
julia> FillArrays.Ones(3) |> typeof |> fieldnames
(:axes,)

julia> FillArrays.One
ERROR: UndefVarError: One not defined

julia> get2(::Val{x}) where x = x^2;

julia> gradient(get2∘Val, 3.14)
(nothing,)

julia> gradient(get2, Val(3.14))
(nothing,)
```

Reply:
Xref weird example (from the tests) here: https://github.com/FluxML/Zygote.jl/pull/1001#discussion_r654866919

tl;dr is that `global s += 1` is not detected by these tests. That's a test of order of iteration.